commit.go

```diff
@@ -107,23 +122,33 @@ func (c *Commit) String() string {
 	)
 }
 
+// CommitIter provides an iterator for a set of commits.
 type CommitIter struct {
 	core.ObjectIter
 	r *Repository
 }
 
+// NewCommitIter returns a CommitIter for the given repository and underlying
+// object iterator.
+//
+// The returned CommitIter will automatically skip over non-commit objects.
 func NewCommitIter(r *Repository, iter core.ObjectIter) *CommitIter {
 	return &CommitIter{iter, r}
 }
 
+// Next moves the iterator to the next commit and returns a pointer to it. If it
+// has reached the end of the set it will return io.EOF.
 func (iter *CommitIter) Next() (*Commit, error) {
-	obj, err := iter.ObjectIter.Next()
-	if err != nil {
-		return nil, err
+	for {
```

Comment:
I added it mostly as a fail-safe, in case someone creates their own ObjectIter that doesn't already filter for the desired type.

Perhaps it would be better instead to create an implementation of ObjectIter that filters for a given type? Then someone could wrap iterators like so:

```
var source ObjectIter // some source of mixed objects
filtered := NewObjectTypeFilter(source, CommitObject)
commits := NewCommitIter(r, filtered)
```

Such a filter could be added in a later commit. In any case, I'm okay with omitting the loop here and in the other iterators if you think it best to keep them simple.